File: core/arch/cpuid.c
/* This file is based on code from the NovariaOS project. 
The source code in NovariaOS is licensed under GPLv3. In this project, the file is available under the MIT license. 
See LICENSE in the root of the repository. 

https://github.com/novariaos/novariaos-src
*/

#include <core/arch/cpuid.h>

void cpuid(uint32_t leaf, uint32_t subleaf, cpuid_result_t* result) {
    asm volatile(
        "mov %4, %%ecx\n\t"
        "cpuid\n\t"
        "mov %%eax, %0\n\t"
        "mov %%ebx, %1\n\t"
        "mov %%ecx, %2\n\t"
        "mov %%edx, %3\n\t"
        : "=m"(result->eax), "=m"(result->ebx), "=m"(result->ecx), "=m"(result->edx)
        : "r"(subleaf), "a"(leaf)
        : "%ebx", "%ecx", "%edx"
    );
}

File: src/arch/cpuid.zig
pub const CpuidResult = struct {
    eax: u32,
    ebx: u32,
    ecx: u32,
    edx: u32,
};

pub fn cpuid(leaf: u32, subleaf: u32) CpuidResult {
    var eax: u32 = undefined;
    var ebx: u32 = undefined;
    var ecx: u32 = undefined;
    var edx: u32 = undefined;
    asm volatile ("cpuid"
        : [eax] "={eax}" (eax),
          [ebx] "={ebx}" (ebx),
          [ecx] "={ecx}" (ecx),
          [edx] "={edx}" (edx),
        : [leaf] "{eax}" (leaf),
          [subleaf] "{ecx}" (subleaf),
    );
    return .{ .eax = eax, .ebx = ebx, .ecx = ecx, .edx = edx };
}

pub fn vendorString(buf: *[12]u8) void {
    const r = cpuid(0, 0);
    const ebx_bytes = @as([4]u8, @bitCast(r.ebx));
    const edx_bytes = @as([4]u8, @bitCast(r.edx));
    const ecx_bytes = @as([4]u8, @bitCast(r.ecx));
    @memcpy(buf[0..4], &ebx_bytes);
    @memcpy(buf[4..8], &edx_bytes);
    @memcpy(buf[8..12], &ecx_bytes);
}

pub fn hasFeature(ecx_bit: u5) bool {
    const r = cpuid(1, 0);
    return (r.ecx >> ecx_bit) & 1 == 1;
}
